dali/imgcodec/image_decoder.cc

```diff
@@ -0,0 +1,614 @@
+                                    ScheduledWork &work,
+                                    const vector<bool> &keep) {
+    int moved = 0;
+
+    int n = work.sources.size();
+
+    for (int i = 0; i < n; i++) {
+      if (keep[i]) {
+        if (moved) {
+          // compact
+          if (!work.cpu_outputs.empty())
+            work.cpu_outputs[i - moved] = std::move(work.cpu_outputs[i]);
+          if (!work.gpu_outputs.empty())
+            work.gpu_outputs[i - moved] = std::move(work.gpu_outputs[i]);
+          if (!work.rois[i])
+            work.rois[i - moved] = std::move(work.rois[i]);
+          work.sources[i - moved] = work.sources[i];
+          work.indices[i - moved] = work.indices[i];
+        }
+      } else {
+        if (fb)
+          fb->move_entry(work, i);
+        moved++;
+      }
+    }
+    if (fb)
+      fb->resize(fb->num_samples() - moved);
+}
+
+void ImageDecoder::ScheduledWork::alloc_temp_cpu_outputs(ImageDecoder &owner) {
+  cpu_outputs.resize(indices.size());
+  for (int i = 0, n = indices.size(); i < n; i++) {
+    int idx = indices[i];
+
+    SampleView<GPUBackend> &gpu_out = gpu_outputs[i];
+
+    // TODO(michalz): Add missing utility functions to SampleView - or just use Tensor again...
+    size_t size = volume(gpu_out.shape()) * TypeTable::GetTypeInfo(gpu_out.type()).size();
+    constexpr int kTempBufferAlignment = 256;
+
+    auto &buf_ptr = owner.temp_buffers_[idx];
+    buf_ptr = mm::alloc_raw_async_shared<char, mm::memory_kind::pinned>(
+      size, mm::host_sync, ctx.stream, kTempBufferAlignment);
+
+    SampleView<CPUBackend> &cpu_out = cpu_outputs[i];
+    cpu_out = SampleView<CPUBackend>(buf_ptr.get(), gpu_out.shape(), gpu_out.type());
+  }
+}
+
+void ImageDecoder::DecoderWorker::process_batch(std::unique_ptr<ScheduledWork> work) noexcept {
+  assert(work->num_samples() > 0);
+  assert((work->cpu_outputs.empty() && work->gpu_outputs.size() == work->sources.size()) ||
+         (work->gpu_outputs.empty() && work->cpu_outputs.size() == work->sources.size()));
+  assert(work->rois.empty() || work->rois.size() == work->sources.size());
+
+  auto mask = decoder_->CanDecode(work->ctx,
+                                  make_span(work->sources),
+                                  work->params,
+                                  make_span(work->rois));
+
+  std::unique_ptr<ScheduledWork> fallback_work;
+  if (fallback_) {
+    fallback_work = owner_->new_work(work->ctx, work->results);
+    move_to_fallback(fallback_work.get(), *work, mask);
+    if (!fallback_work->empty())
+      fallback_->add_work(std::move(fallback_work));
+  } else {
+    filter(*work, mask);
+    for (size_t i = 0; i < mask.size(); i++) {
+      if (!mask[i])
+        work->results.set(work->indices[i], DecodeResult::Failure(nullptr));
+    }
+  }
+
+  if (!work->sources.empty()) {
+    bool decode_to_gpu = produces_gpu_output_;
+
+    if (!decode_to_gpu && work->cpu_outputs.empty()) {
```

Comment:
are we sure that pre-existing `cpu_outputs` have the right shape?

Reply:
Of course they don't because there aren't any.